pkg/mev/bundle_sender.go

```diff
@@ -124,6 +134,35 @@ func (s *Client) flashbotBackrunSendBundle(
 	return res, err
 }
 
+func (s *Client) EstimateBundleGas(
+	_ context.Context,
+	messages []ethereum.CallMsg,
+	overrides *map[common.Address]gethclient.OverrideAccount,
+) ([]uint64, error) {
+	bundles := make([]interface{}, 0, len(messages))
+	for _, msg := range messages {
+		bundles = append(bundles, ToCallArg(msg))
+	}
+
+	var gasEstimateCost []hexutil.Uint64
+
+	err := s.ethClient.Client().Call(
+		&gasEstimateCost, ETHEstimateGasBundleMethod,
+		map[string]interface{}{
+			"transactions": bundles,
+		}, "latest", overrides,
+	)
```

Comment:
Estimate gas for bundle is not supported by node, why need to call to `ethClient` to estimate gas for bundle? Why don't use a builder client to estimate it?

Reply:
Using estimate gas for the bundle by builder client requires signed txs (using `eth_callBundle`). But the current flow requires we estimate gas before signing Tx (https://github.com/KyberNetwork/poseidon/blob/23075396687934c69eaa21507577255170220562/internal/executor/blockchain/blockchain.go#L959)

So if I estimate gas for backrun tx, I might got return amount is not enough. And then I can't have a signed tx to estimate gas bundle.

So I decided to use our trading node's custom estimate gas bundle rpc to estimate gas bundle without signed txs (only need `ethereum.CallMsg`).

Pls correct me if have another way

Reply:
So you should define an interface for estimating a bundle with overrides. I'm also thinking about a way to use `eth_estimateGas` with overrides pools' state for estimating gas for backrun tx.

Reply:
```
type IBundleSender interface {
	.....
	EstimateBundleGas(
		ctx context.Context,
		messages []ethereum.CallMsg,
		overrides *map[common.Address]gethclient.OverrideAccount,
	) ([]uint64, error)
	....
}
```

I have defined the `EstimateBundleGas` func signature in the IBundleSender interface. You mean I should define another new interface for estimating a bundle right?

Reply:
Thanks. I've defined a new interface and refactor by following this way